src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.OpenFileTracker.cs

```diff
@@ -108,72 +138,74 @@ void IRunningDocumentTableEventListener.OnRenameDocument(string newMoniker, stri
 
-                    var textContainer = textBuffer.AsTextContainer();
-
-                    foreach (var documentId in documentIds)
-                    {
-                        if (!w.IsDocumentOpen(documentId) && !_workspace._documentsNotFromFiles.Contains(documentId))
+                        foreach (var documentId in documentIds)
                         {
-                            var isCurrentContext = documentId.ProjectId == activeContextProjectId;
-                            if (w.CurrentSolution.ContainsDocument(documentId))
-                            {
-                                w.OnDocumentOpened(documentId, textContainer, isCurrentContext);
-                            }
-                            else if (w.CurrentSolution.ContainsAdditionalDocument(documentId))
-                            {
-                                w.OnAdditionalDocumentOpened(documentId, textContainer, isCurrentContext);
-                            }
-                            else
+                            if (!w.IsDocumentOpen(documentId) && !_workspace._documentsNotFromFiles.Contains(documentId))
                             {
-                                Debug.Assert(w.CurrentSolution.ContainsAnalyzerConfigDocument(documentId));
-                                w.OnAnalyzerConfigDocumentOpened(documentId, textContainer, isCurrentContext);
+                                var isCurrentContext = documentId.ProjectId == activeContextProjectId;
+                                if (w.CurrentSolution.ContainsDocument(documentId))
+                                {
+                                    w.OnDocumentOpened(documentId, textContainer, isCurrentContext);
+                                }
+                                else if (w.CurrentSolution.ContainsAdditionalDocument(documentId))
+                                {
+                                    w.OnAdditionalDocumentOpened(documentId, textContainer, isCurrentContext);
+                                }
+                                else
+                                {
+                                    Debug.Assert(w.CurrentSolution.ContainsAnalyzerConfigDocument(documentId));
+                                    w.OnAnalyzerConfigDocumentOpened(documentId, textContainer, isCurrentContext);
+                                }
                             }
                         }
-                    }
+                    }).ConfigureAwait(false);
                 });
             }
 
-            private ProjectId GetActiveContextProjectIdAndWatchHierarchies_NoLock(string moniker, IEnumerable<ProjectId> projectIds, IVsHierarchy? hierarchy)
+            private async Task<ProjectId> GetActiveContextProjectIdAndWatchHierarchiesAsync(string moniker, IEnumerable<ProjectId> projectIds, IVsHierarchy? hierarchy)
             {
+                await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync();
```

Comment:
I have to put the comment here, but note that later there is:

```
GetProjectWithHierarchyAndName_NoLock
```

Which is assuming this code is running while acquiring the workspace lock. The loss of _NoLock suffix here was what somewhat tipped me off.

Reply:
yup.  good catch!